hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -1339,7 +1339,38 @@ public void bootstrap(OMNodeDetails newOMNode) throws IOException {
   }
 
   /**
-   * Add a new OM Node to the HA cluster. This call comes from OMRatisServer
+   * When OMStateMachine receives a configuration change update, it calls
+   * this function to update the peers list, if required.
+   */
+  public void updatePeerList(List<String> omNodeIds) {
+    List<String> ratisServerPeerIdsList = omRatisServer.getPeerIds();
+    for (String omNodeId : omNodeIds) {
+      if (getOMNodeId().equals(omNodeId)) {
```

Comment:
In normal OM start node, ratisServer peer list has local + peer.

```
 if (!isBootstrapping) {
      // On regular startup, add all OMs to Ratis ring
      raftPeers.add(localRaftPeer);
```
In bootstrap case, local node is not added to ratis server peer list.

Reply:
While reading more, I have observed that before this change, we have not added local OM node to peer node  list before(raftPeers). Should we have the same behavior?

Reply:
> In bootstrap case, local node is not added to ratis server peer list.

Yes. This is because the local node is not part of the ratis ring yet. When the SetConfiguration is executed, this node would be added to the peer list.

> While reading more, I have observed that before this change, we have not added local OM node to peer node list before(raftPeers). Should we have the same behavior?. So, should we remove adding local node to peer list?

Before this change also the local OM node was added to the peer list in OMRatisServer. Are you referring to any other class?

Reply:
```
newOMRatisServer
    for (OMNodeDetails peerInfo : peerNodes) {
      String peerNodeId = peerInfo.getNodeId();
      RaftPeerId raftPeerId = RaftPeerId.valueOf(peerNodeId);
      RaftPeer raftPeer;
      if (peerInfo.isHostUnresolved()) {
        raftPeer = RaftPeer.newBuilder()
            .setId(raftPeerId)
            .setAddress(peerInfo.getRatisHostPortStr())
            .build();
      } else {
        InetSocketAddress peerRatisAddr = new InetSocketAddress(
            peerInfo.getInetAddress(), peerInfo.getRatisPort());
        raftPeer = RaftPeer.newBuilder()
            .setId(raftPeerId)
            .setAddress(peerRatisAddr)
            .build();
      }

      // Add other OM nodes belonging to the same OM service to the Ratis ring
      raftPeers.add(raftPeer);
    }

    return new OzoneManagerRatisServer(ozoneConf, omProtocol, omServiceId,
        localRaftPeerId, ratisAddr, raftPeers, secConfig, certClient);

Looks like previously we don't save raftPeers in OzoneManagerRatisServer.

But now with this PR, on old node all nodes peer+local will be in peer list. 

if (!isBootstrapping) { 
// On regular startup, add all OMs to Ratis ring 
raftPeers.add(localRaftPeer);

Where as for bootstrap due to this skip, we don't add

if (getOMNodeId().equals(omNodeId)) {
continue;
}
Due to this on bootstrap node peerlist does not include local node

There is a difference in raftPeers list in bootstrap and old node.

Reply:
> There is a difference in raftPeers list in bootstrap and old node.

Correct. On a bootstrapping node, we do not add the local node to raftPeer list. This is done later when a setConfiguration request is executed and it calls OMRatisServer#addRaftPeer().

```
  /**
   * Add given node to list of RaftPeers.
   */
  public void addRaftPeer(OMNodeDetails omNodeDetails) {
    InetSocketAddress newOMRatisAddr = new InetSocketAddress(
        omNodeDetails.getHostAddress(), omNodeDetails.getRatisPort());

    raftPeers.add(RaftPeer.newBuilder()
        .setId(RaftPeerId.valueOf(omNodeDetails.getNodeId()))
        .setAddress(newOMRatisAddr)
        .build());

    LOG.info("Added OM {} to Ratis Peers list.", omNodeDetails.getNodeId());
  }
```

As for the old code, the local node was added to raftPeers there too (OzoneManagerRatisServer Line#348 in old code).

Reply:
This is done later when a setConfiguration request is executed and it calls OMRatisServer#addRaftPeer().

Yes, but notifyConfigurationChanged calls updatePeerList

And in updatePeerList we have 

    for (String omNodeId : omNodeIds) {
```
      if (getOMNodeId().equals(omNodeId)) {
        continue;
      }
```

Which skips calling addRaftPeer, when it is matching with localNode ID

Reply:
Thanks Bharat. I get it now. Should not skip the checks for local node. I have updated the patch. Thanks for catching this.